### PR TITLE
Add workflow that triggers release on due date

### PR DESCRIPTION
**Pull request by eleftherias**
Add Gradle task that calculates the next release milestone based on the current version and check if it is due today.

The second step in the workflow is only a placeholder.

Issue gh-10451
Issue gh-10455

### CODE REVIEW

File: buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestoneNextVersionDueTodayTask.java
Comment (sjohnr) [inline buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestoneNextVersionDueTodayTask.java:55]: After discussion, I did a quick test, and it looks like if we only output the next version if `milestoneDueToday` is true, and (possibly) a blank line otherwise, we can capture that output with the `-q` flag.

For example:

```sh
nextVersion=$(./gradlew -q gitHubCheckMilestoneDueToday)
if [[ "$nextVersion" != "" ]]
then
    echo "Starting release for next version"
fi
```

**Note:** We could also set a task output and check that in a task further down, if we want to avoid the bash if-statement.

Perhaps we could split this into two tasks:

1. `gitHubCheckMilestoneDueToday` fails the build if the given/calculated version is not due today (can be used as a guard against accidentally releasing at the wrong time)
2. `gitHubCalculateNextReleaseMilestoneForToday` (or similar) which simply outputs the next version, or nothing/blank line

File: .github/workflows/release-next-version.yml
Comment (sjohnr) [inline .github/workflows/release-next-version.yml:72]: As mentioned below, I feel like capturing the next version (either in a variable or a task output) really unlocks the potential for what we would do in this placeholder task.

**Review by sjohnr — COMMENTED** [review]
This looks great! See discussion comments below.

File: buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestoneNextVersionDueTodayTask.java
Comment (eleftherias) [reply buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestoneNextVersionDueTodayTask.java:55]: I agree it makes sense to split it up into 2 tasks. I was just talking to Rob and he suggested looking into output / input files, since that is a standard way to share information between tasks. I'll start with these updates over the next few days.

Comment (eleftherias) [reply buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestoneNextVersionDueTodayTask.java]: As you suggested I split it up into two tasks:
1. `gitHubNextReleaseMilestone` which outputs the next version to a file called `next-release.yml`
2. `gitHubCheckMilestoneDueToday` which depends on `gitHubNextReleaseMilestone` and uses the generated file `next-release.yml` as input and fails the build if the given/calculated version is not due today

File: .github/workflows/release-next-version.yml
Comment (eleftherias) [reply .github/workflows/release-next-version.yml]: The task that performs the release could depend on the task `gitHubNextReleaseMilestone`, similar to how `gitHubCheckMilestoneDueToday` does it.

File: .github/workflows/release-next-version.yml
Comment (sjohnr) [inline .github/workflows/release-next-version.yml:6]: I believe the [`schedule` trigger](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) runs only on the default branch. I'm thinking we may need a higher level workflow to trigger these builds on each branch.

Should we solve that as part of this PR or go ahead and merge this and work on that separately? If separately, perhaps remove the schedule and only include the manual trigger here?

I personally think the manual trigger adds value on other branches even without a scheduled run!

**Review by sjohnr — APPROVED** [review]
Thanks @eleftherias! I've tested this and it works very well! See one comment below, but beyond that this seems ready to merge and add to the 5.7 branch as well.

File: .github/workflows/release-next-version.yml
Comment (eleftherias) [reply .github/workflows/release-next-version.yml:6]: At the moment the only way to run a scheduled workflow on the 5.7.x branch is to add another workflow, for example `release-next-version-5.7.x.yml`, to the main branch. This isn't ideal since the CI code won't be tied to the source code, they'll be on different branches.
The `release-next-version-5.7.x.yml` would look like this:
```
# ...
jobs:
  # ...
  release:
    name: Release next version
    needs: [check_release_due]
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
        with:
          ref: 5.7.x
# ...
```
I would like to hold off on adding this until we have a complete release pipeline, in case GitHub has added some new functionality by then.

The relevant GitHub Actions issue is [here](https://github.community/t/scheduled-builds-of-non-default-branch/16306).

Comment (sjohnr) [reply .github/workflows/release-next-version.yml:6]: Thanks! I agree, let's hold off. Just wanted to note it for future me/you/anyone else.

**Pull request merged**
